File: ruby/perfect-numbers/perfect_numbers.rb
class PerfectNumber
  def self.classify(number)
    raise RuntimeError if number <= 1

    new(number).classify
  end

  def initialize(number)
    @number = number
  end

  def classify
    case factors.sum <=> number
    when -1
      'deficient'
    when 0
      'perfect'
    when 1
      'abundant'
    end
  end

  def factors
    (1...number).filter { |n| (number % n).zero? }
  end

  private

  attr_reader :number
end

File: haskell/perfect-numbers/src/PerfectNumbers.hs
module PerfectNumbers
  ( classify
  , Classification(..)
  )
where

data Classification = Deficient | Perfect | Abundant deriving (Eq, Show)

classify :: Int -> Maybe Classification

classify n
  | n <= 0
  = Nothing
  | otherwise
  = let factors = [ i | i <- [1 .. n `div` 2], n `rem` i == 0 ]
    in  case sum factors `compare` n of
          LT -> Just Deficient
          EQ -> Just Perfect
          GT -> Just Abundant

